File: python/table.py
import math
import constants
from pager import Pager
from leaf_node import Leaf_Node

class Table:

  def __init__(self, filename):
    self.pager = Pager(filename)
    self.root_page_num = 0

    # If this is a new database file: initialize the root node
    if self.pager.num_pages == 0:
      root_node = Leaf_Node(self.pager.get_page(self.root_page_num))
      root_node.is_root = True
      self.pager.insert_new_page(root_node)


  def get_page(self, page_num):
    return self.pager.get_page(page_num)

  def get_root_page(self):
    return self.get_page(self.root_page_num)

  # save pages to the file
  def flush(self):
    for page_num in range(0, self.pager.num_pages):
      self.pager.flush_page(page_num)

    self.pager.file.close()

File: c/table.c
#include <fcntl.h>
#include <errno.h>
#include <unistd.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "table.h"

Table* table_open(const char* filename) {
  Pager* pager = pager_open(filename);
  Table* table = malloc(sizeof(Table));

  table->pager = pager;
  table->root_page_num = 0;

  if (pager->num_pages == 0) {
    // New database file. Initialize page 0 as leaf node.
    void* root_node = get_page(pager, 0);
    initialize_leaf_node(root_node);
    set_node_root(root_node, true);
  }
 
  return table;
}

void free_table(Table* table) {
  for (int i = 0; table->pager->pages[i]; i++) {
    free(table->pager->pages[i]);
  }
  free(table);
}

void table_flush(Table* table) {
  Pager* pager = table->pager;

  for (uint32_t i = 0; i < pager->num_pages; i++) {
    if (pager->pages[i] == NULL) {
      continue;
    }
    pager_flush(pager, i);
    free(pager->pages[i]);
    pager->pages[i] = NULL;
  }
}

void table_deallocate(Table* table) {
  Pager* pager = table->pager;

  for (uint32_t i = 0; i < TABLE_MAX_PAGES; i++) {
    void* page = pager->pages[i];
    if (page) {
      free(page);
      pager->pages[i] = NULL;
    }
  }

  free(pager);
  free(table);
}


void table_close(Table* table) {
  Pager* pager = table->pager;

  table_flush(table);

  // close the file
  int result = close(pager->file_descriptor);
  if (result == -1) {
    printf("Error closing db file.\n");
    exit(EXIT_FAILURE);
  }

  table_deallocate(table);
}
